README
Math::Random::ISAAC::XS is a C implementation of the ISAAC PRNG Algorithm

Copyright 2011 by Jonathan Yu <jawnsy@cpan.org>. All rights reserved.

INSTALLATION

  This module depends on Module::Build for its installation mechanism.

  To install this module type the following:
  $ perl Build.PL
  $ perl Build
  $ perl Build test
  $ perl Build install

  Module::Build will complain if any dependencies are missing, or if
  you can install some other modules to enhance its functionality.

USAGE AND DOCUMENTATION

  In order to use the module, you'll probably need the documentation. Once
  installed, you can read that using:
  $ perldoc Math::Random::ISAAC::XS

LICENSE

  In a perfect world, I could just say that this package and all of the
  code it contains is Public Domain. It's a bit more complicated than
  that; you'll have to read LICENSE to get the full details.

  Here are some nice things to live by:
  1. Give credit where it's due.
  2. Share your modifications, so everyone benefits from them.
  3. Do something nice for the authors.
  4. Help someone who needs it.
  5. Don't waste. Anything. Be especially careful not to waste energy that
     comes from non-renewable sources.
  6. Be tolerant. Everything that's good in nature comes from cooperation.
  Paraphrased a bit from Alberto Bertogli's Buena Onda License Agreement.
  For the full text, see: http://blitiri.com.ar/p/bola/


Changes
{{$NEXT}}

1.004     2011-02-17 22:04:28 America/Toronto

          Correct dependencies to indicate Math::Random::ISAAC
          is a runtime dependency (thanks rjbs)

          Lower Module::Build requirement to 0.2808_01, we successfully
          built with this version for 1.001.

1.003     2011-02-13 14:54:07 America/Toronto

          Migrate package to Dist::Zilla format

1.002     2011-02-13

          Changed my mail address to jawnsy@cpan.org

          Update POD

          Update MANIFEST.SKIP according to new defaults provided by
          ExtUtils::Manifest

          Remove pass-through Makefile.PL

1.001     2009-08-18

          Remove generated files from the repository

          Use a simple string version, per David Golden's article:
          http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/

          No longer use Perl::Critic to test

          Remove modules available in core since Perl 5.6 from requires

1.0.6     2009-07-25

          Rewrote POD a bit

          The last variables using TEST_AUTHOR now use
          RELEASE_TESTING|AUTOMATED_TESTING

          Overhaul test scripts, remove author tests from Recommends
          (Alias' journal #5375)

          Provide a more generic README so I don't need to bother
          updating it.

          Minor cleanups in the LICENSE

          Removed CAVEATS from README

          Updated copyright statements

          Require Internet (TEST_INTERNET) to do the signature test,
          since it tries to download the key from a keyserver

          Slight reformatting, nothing significant.

          No longer enforce "one dot only" in filenames. This
          breaks compatibility with OpenVMS.

          Minor formatting changes

          Instead of joining the array, just pass an array ref
          (otherwise it is split into an array anyway) [Build.PL]

          Clarified license for the case where public domain is
          not permitted (ie, Germany)

          Applied patch from RT#46071 -- thanks to Taro Nishino.
          See: http://rt.cpan.org/Ticket/Display.html?id=46071

1.0.5     2009-05-12

          Add the Perl portability header ppport.h to MANIFEST

          New simplified license - Public Domain (fallback on MIT,
          Artistic-2+, GPL-2+ only)

          Removed requirement for minimum 'version' version

          Added ppport.h for portability to older Perls (this
          hasn't been a problem so far, but `perl ppport.h'
          suggested it due to SvUV)

          Added ExtUtils::ParseXS as a dependency; modules were
          failing to build on systems that don't have it installed,
          because the typemap isn't properly included by
          xsubpp/Module::Build. See:
          http://osdir.com/ml/lang.perl.modules.module-build/2005/msg00501.html

          Made some changes mostly related to splint warnings:
          o Declared isaac() as static because it isn't used
            outside rand.c

          Minor reformatting

          Changed line to: if (ctx->randcnt-- == 0) { to make
          intent more obvious

1.0.4     2009-05-06

          Changed the valgrind test from an author test to its own
          special test - one must set TEST_VALGRIND=1 to test the
          module through valgrind. This must be done until we have
          better suppression support upstream in Test::Valgrind.

          Updated t/02valgrind.t so that tests are not planned
          (so Test::Valgrind can plan)

          Removed version-specific stuff from the README

          Reformatted the POD file

          Updated support information in META.yml and Build.PL

          Added some support information

          Reformatted rand.c a bit; moved variable definitions
          to the top

          Added .tmp to MANIFEST.SKIP

          Made this a dynamic config (instead of using META.yml)

          Added ExtUtils::CBuilder to META.yml

          Added Build.bat to MANIFEST.SKIP for Win32 builds...

1.0.3     2009-04-23

          Rewrote docs a bit

          Removed Pod::Usage dependency

          Added some example files for benchmarking the module and
          charting out the distribution generated by the module :-)

          Added Config to META.yml prerequisites (why not?). It's a
          core module so it wouldn't be missing anyway, and Build.PL
          would fail outright if it was.

1.0.2     2009-04-21

          Add Build.PL detection of int size

          Version bumped to 1.0.2

          A flag now controls whether 'int' or 'long' is used as the
          type: USE_INT. If Build.PL detects that int is 32bits wide,
          then it will use that type. Otherwise it'll use long, which
          is more portable.

          A flag will control if the portability code is used. Right
          now this involves masking all numbers with "& 0xffffffff"
          to cut off the high bits - this code isn't necessary for
          32-bit variable widths, so we might squeeze a bit more
          performance out this way

1.0.1     2009-04-07

          Changed the bootstrap loading to use VERSION instead
          of VERSION->numify, so that they match (prevents compile
          errors with "object version v1.0 does not match bootstrap
          parameter 1.000000" (apparently causes failures in 50% of
          machines so far)

1.0       2009-04-06

          Initial release


lib/Math/Random/ISAAC/XS.pm
package Math::Random::ISAAC::XS;
# ABSTRACT: C implementation of the ISAAC PRNG algorithm

use strict;
use warnings;

=head1 SYNOPSIS

This module implements the same interface as C<Math::Random::ISAAC> and can
be used as a drop-in replacement. This is the recommended implementation of
the module, based on Bob Jenkins' reference implementation in C.

Selecting the backend to use manually really only has two uses:

=over

=item *

If you are trying to avoid the small overhead incurred with dispatching
method calls to the appropriate backend modules.

=item *

If you are testing the module for performance and wish to explicitly decide
which module you would like to use.

=back

Example code:

  # With Math::Random::ISAAC
  my $rng = Math::Random::ISAAC->new(time);
  my $rand = $rng->rand();

  # With Math::Random::ISAAC::XS
  my $rng = Math::Random::ISAAC::XS->new(time);
  my $rand = $rng->rand();

=head1 DESCRIPTION

See L<Math::Random::ISAAC> for the full description.

=head1 METHODS

=head2 new

  Math::Random::ISAAC::XS->new( @seeds )

Implements the interface as specified in C<Math::Random::ISAAC>

=head2 rand

  $rng->rand()

Implements the interface as specified in C<Math::Random::ISAAC>

=head2 irand

  $rng->irand()

Implements the interface as specified in C<Math::Random::ISAAC>

=cut

# This is the code that actually bootstraps the module and exposes
# the interface for the user. XSLoader is believed to be more
# memory efficient than DynaLoader.
use XSLoader;
XSLoader::load(__PACKAGE__, $Math::Random::ISAAC::XS::VERSION);

=head1 SEE ALSO

L<Math::Random::ISAAC>

1;


Build.PL
#!/usr/bin/perl

# Build.PL
#  Script to build and install this distribution

use strict;
use warnings;
use Config;

use Module::Build;

my @flags;

# Determine what the 32bit unsigned type is. Based on the result, we can
# pass two flags to the C code which will affect their behaviour.
#
# USE_PORTABLE will determine whether we try to make our bit operations
# portable by ANDing them with 0xffffffff. This cuts off higher bits,
# so that our result is good.
#
# USE_INT will determine whether to use the 'int' type compared to 'long'.
# While 'long' is guaranteed to be 32 bits or larger, on some 64-bit systems,
# int is 32 bits and long is 64 bits. On these systems we force use of int.
if ($Config{intsize} == 4) {
  # Excellent, we've got a 32-bit type. It probably doesn't matter whether we
  # use int or long here, if they are the same size. However, they would be
  # just as fast anyway, I think.
  push(@flags, '-DUSE_INT');
}
else {
  # Use the 'portable' version and hope for the best. It doesn't seem to work
  # very well right now, hopefully it's fixed for fully-64bit architectures
  # later.
  push(@flags, '-DUSE_PORTABLE');
}

# Note that even though they are currently presented that way, this does not
# necessarily mean that we must either USE_INT or USE_PORTABLE. There may be
# real cases where we don't want to USE_INT (and instead use long's, if they
# are somehow faster than ints) but are absolutely sure they are only 32bits.
# In that case we wouldn't need to USE_PORTABLE either.

my $builder = Module::Build->new(
  module_name           => 'Math::Random::ISAAC::XS',
  license               => 'unrestricted',
  dist_version_from     => 'lib/Math/Random/ISAAC/XS.pm',
  dynamic_config        => 1,

  # Location of our special C and XS source files
  c_source => 'src',
  xs_files => {
    'src/ISAAC.xs' => 'lib/Math/Random/ISAAC/XS.xs',
  },

  extra_compiler_flags => \@flags,

  # Prerequisites generated from dist.ini by ModuleBuild::Custom
##{ $plugin->get_prereqs ##}

  add_to_cleanup => [ 'Math-Random-ISAAC-XS-*' ],
);

$builder->create_build_script();
